compiler/src/dotty/tools/dotc/core/GadtConstraint.scala

```diff
@@ -71,16 +72,31 @@ class GadtConstraint private (
     externalize(constraint.nonParamBounds(param)).bounds
 
   def fullLowerBound(param: TypeParamRef)(using Context): Type =
-    constraint.minLower(param).foldLeft(nonParamBounds(param).lo) {
-      (t, u) => t | externalize(u)
+    val self = externalize(param)
+    constraint.minLower(param).foldLeft(nonParamBounds(param).lo) { (acc, p) =>
+      externalize(p) match
+        case tp: TypeRef
+          // drop any lower param that is a GADT symbol
+          // and is upper-bounded by a non-Any super-type of the original parameter
+          // e.g. in pos/i14287.min
+          // B$1 had info <: X   and fullBounds >: B$2 <: X, and
+          // B$2 had info <: B$1 and fullBounds <: B$1
+          // We can use the info of B$2 to drop the lower-bound of B$1
+          // and return non-bidirectional bounds B$1 <: X and B$2 <: B$1.
+          if tp.name.is(UniqueName) && !tp.info.hiBound.isExactlyAny && self <:< tp.info.hiBound => acc
```

Comment:
Can we use `tp.symbol.isPatternBound` instead of relying on the name kind?

Reply:
Also, would checking `self =:= tp.info.hiBound` be enough for all our test cases? That would remove the need for `isExactlyAny` which I'm a bit wary of (what if the symbols involved have another common upper-bound like AnyRef?)